src/common/providers/https.ts

```diff
@@ -0,0 +1,581 @@
+    const obj: Record<string, any> = {};
+    for (const [k, v] of Object.entries(data)) {
+      obj[k] = decode(v);
+    }
+    return obj;
+  }
+  // Anything else is safe to return.
+  return data;
+}
+
+/**
+ * Be careful when changing token status values.
+ *
+ * Users are encouraged to setup log-based metric based on these values, and
+ * changing their values may cause their metrics to break.
+ *
+ */
+type TokenStatus = 'MISSING' | 'VALID' | 'INVALID';
+
+interface CallableTokenStatus {
+  app: TokenStatus;
+  auth: TokenStatus;
+}
+
+/**
+ * Check and verify tokens included in the requests. Once verified, tokens
+ * are injected into the callable context.
+ *
+ * @param {Request} req - Request sent to the Callable function.
+ * @param {CallableContext} ctx - Context to be sent to callable function handler.
+ * @return {CallableTokenStatus} Status of the token verifications.
+ */
+async function checkTokens(
+  req: Request,
+  ctx: CallableContext
+): Promise<CallableTokenStatus> {
+  const verifications: CallableTokenStatus = {
+    app: 'MISSING',
+    auth: 'MISSING',
+  };
+
+  const appCheck = req.header('X-Firebase-AppCheck');
+  if (appCheck) {
+    verifications.app = 'INVALID';
+    try {
+      if (!apps().admin.appCheck) {
+        throw new Error(
+          'Cannot validate AppCheck token. Please update Firebase Admin SDK to >= v9.8.0'
+        );
+      }
+      const appCheckToken = await apps()
+        .admin.appCheck()
+        .verifyToken(appCheck);
+      ctx.app = {
+        appId: appCheckToken.appId,
+        token: appCheckToken.token,
+      };
+      verifications.app = 'VALID';
+    } catch (err) {
+      logger.warn('Failed to validate AppCheck token.', err);
+    }
+  }
+
+  const authorization = req.header('Authorization');
+  if (authorization) {
+    verifications.auth = 'INVALID';
+    const match = authorization.match(/^Bearer (.*)$/);
+    if (match) {
+      const idToken = match[1];
+      try {
+        const authToken = await apps()
+          .admin.auth()
+          .verifyIdToken(idToken);
+
+        verifications.auth = 'VALID';
+        ctx.auth = {
+          uid: authToken.uid,
+          token: authToken,
+        };
+      } catch (err) {
+        logger.warn('Failed to validate auth token.', err);
+      }
+    }
+  }
+
+  const logPayload = {
+    verifications,
+    'logging.googleapis.com/labels': {
+      'firebase-log-type': 'callable-request-verification',
+    },
+  };
+
+  const errs = [];
+  if (verifications.app === 'INVALID') {
+    errs.push('AppCheck token was rejected.');
+  }
+  if (verifications.auth === 'INVALID') {
+    errs.push('Auth token was rejected.');
+  }
+
+  if (errs.length == 0) {
+    logger.info('Callable request verification passed', logPayload);
+  } else {
+    logger.warn(
+      `Callable request verification failed: ${errs.join(' ')}`,
+      logPayload
+    );
+  }
+
+  return verifications;
+}
+
+/** @internal */
+export let onCallHandler = (
```

Comment:
Is there a reason for `export let` vs. `export const`? Also why not just `export function`?

Lastly, I wonder if this should be a factory instead `onCallHandler({cors: Corsoptions})`, that way a critical piece of functionality (making the request pass cors) can't be accidentally forgotten in a future version.